src/Language/Haskell/TH/Datatype.hs

```diff
@@ -1804,8 +1833,17 @@ instance TypeSubstitution Type where
       AppKindT t k  -> freeVariables t `union` freeVariables k
       ImplicitParamT _ t
                     -> freeVariables t
+#endif
+#if MIN_VERSION_template_haskell(2,16,0)
+      ForallVisT tvs t'
+                    -> fvs_under_forall tvs (freeVariables t')
 #endif
       _             -> []
+    where
+      fvs_under_forall :: [TyVarBndr] -> [Name] -> [Name]
+      fvs_under_forall tvs fvs =
+        (concatMap (freeVariables . tvKind) tvs `union` fvs)
+        \\ map tvName tvs
```

Comment:
Is there a concern here that a duplicate variable would be missed?

Consider:

```
>>> "aa" `union` [] \\ "a"
"a"
```

Reply:
Eep! Good catch. Here is a concrete test case:

```
λ> -- forall k (a :: k) (b :: k). ()
λ> freeVariables $ ForallT [PlainTV (mkName "k"), KindedTV (mkName "a") (VarT (mkName "k")), KindedTV (mkName "b") (VarT (mkName "k"))] [] (ConT ''())
[k]
```

I suspect the use of `concatMap` there is wrong. Fix incoming.

Reply:
I've just pushed a change which adds a `t80Test` regression test for this example. Hooray for the `TypeSubstitution [a]` instance!

As far as the use of `\\` itself goes, I think it is OK as long as `union` always deletes duplicate occurrences of free variables. The previous bug occurred because we were not using `union` where we needed to, after all.